master/buildbot/test/integration/test_locks.py

```diff
@@ -120,11 +142,292 @@ def test_builder_lock_release_wakes_builds_for_another_builder(self):
+
+        yield self.assert_two_builds_created_one_after_another(
+            stepcontrollers, master, builder_ids)
+
+
+class TestReconfig(RunFakeMasterTestCase):
+
+    def create_stepcontrollers(self, count, lock, mode):
+        stepcontrollers = []
+        for i in range(count):
+            locks = [lock.access(mode)] if lock is not None else []
+            stepcontrollers.append(BuildStepController(locks=locks))
+        return stepcontrollers
+
+    def update_builder_config(self, config_dict, stepcontrollers, lock, mode):
+        config_dict['builders'] = []
+        for i, stepcontroller in enumerate(stepcontrollers):
+            locks = [lock.access(mode)] if lock is not None else []
+            b = BuilderConfig(name='builder{}'.format(i),
+                              workernames=['worker1'],
+                              factory=BuildFactory([stepcontroller.step]),
+                              locks=locks)
+            config_dict['builders'].append(b)
+
+    @defer.inlineCallbacks
+    def create_single_worker_n_builder_lock_config(self, builder_count,
+                                                   lock_cls, max_count, mode):
+        stepcontrollers = self.create_stepcontrollers(builder_count, None, None)
+
+        lock = lock_cls("lock1", maxCount=max_count)
+
+        config_dict = {
+            'builders': [],
+            'workers': [
+                self.createLocalWorker('worker1'),
+            ],
+            'protocols': {'null': {}},
+            'multiMaster': True,
+        }
+        self.update_builder_config(config_dict, stepcontrollers, lock, mode)
+
+        master = yield self.getMaster(config_dict)
+
+        builder_ids = []
+        for i in range(builder_count):
+            builder_ids.append((
+                yield master.data.updates.findBuilderId('builder{}'.format(i))))
+
+        return stepcontrollers, master, config_dict, lock, builder_ids
+
+    @defer.inlineCallbacks
+    def create_single_worker_n_builder_step_lock_config(self, builder_count,
+                                                        lock_cls, max_count,
+                                                        mode):
+        lock = lock_cls("lock1", maxCount=max_count)
+        stepcontrollers = self.create_stepcontrollers(builder_count, lock, mode)
+
+        config_dict = {
+            'builders': [],
+            'workers': [
+                self.createLocalWorker('worker1'),
+            ],
+            'protocols': {'null': {}},
+            'multiMaster': True,
+        }
+        self.update_builder_config(config_dict, stepcontrollers, None, None)
+
+        master = yield self.getMaster(config_dict)
+
+        builder_ids = []
+        for i in range(builder_count):
+            builder_ids.append((
+                yield master.data.updates.findBuilderId('builder{}'.format(i))))
+
+        return stepcontrollers, master, config_dict, lock, builder_ids
+
+    @parameterized.expand([
```

Comment:
this looks like lots of new integration tests. Wouldn't that raise the test time too much? Normally, you would not test lots of different cases in integration, but rather do it in unit testing

Reply:
Normally I would agree, but in this case I think that the benefits are worth the overhead. The tested functionality is a core part of buildbot and is also prone to subtle issues. Each of the different cases test different interaction of various components and that interaction is hard to properly cover by unit tests. I believe that given how time consuming is investigation of lock-related issues, if these integration tests catch even a single bug then they would still pay off a year of their overhead for the entire buildbot development team. Furthermore, as time goes by and computing resources become cheaper, the cost of these tests will become relatively smaller and smaller.

On my machine, the new tests slowed down test execution by around 2-4% when running tests on single core.

On buildbot test farm the current test with longest runtime is `pylint` by a wide margin (around 2x), so the new tests should not affect the total latency of the entire test suite.

Reply:
ok

Reply:
Thanks!